### PR TITLE
Conditionals 6: NumPy assignments with Vulcan conditionals

### PR DESCRIPTION
**Pull request by havogt**

### CODE REVIEW

**Review by egparedes — COMMENTED** [review]
I fixed some format issues and expanded a bit the examples. I also have a couple of questions about the conditionals. Other than that, it looks fine to me.

File: GTScript-Parallel-model.md [hunk 1]
```diff
@@ -134,17 +186,115 @@ The computation domain of every statement is extended to ensure that any require
...
+    parfor ij:
+        if not my_config_var:
+            b[i, j, k] = 1
+```
+
+### Conditionals on field expressions
+
+- The condition is evaluated for all gridpoints and stored in a mask.
+- Each statement inside the if and else branches is executed according to the same rules as statements outside of branches.
+- Inside the if and else blocks the same field cannot be written to **and** read with an offset (order does not matter).
```
[the remark this answers is not in the document]
Comment (havogt) [reply GTScript-Parallel-model.md]: - limitation is only on the parallel axises as you said
- self-assignments with offset are not possible as they are reads with offset + a write

Updated accordingly/

File: GTScript-Parallel-model.md [hunk 2]
```diff
@@ -134,17 +186,119 @@ The computation domain of every statement is extended to ensure that any require
...
+    parfor ij:
+        if not my_config_var:
+            b[i, j, k] = 1
+```
+
+### Conditionals on field expressions
+
+- The condition is evaluated for all gridpoints and stored in a mask.
+- Each statement inside the if and else branches is executed according to the same rules as statements outside of branches.
+- Inside the if and else blocks the same field cannot be written to **and** read with an offset in the parallel axises (order does not matter).
```
Comment (egparedes) [inline GTScript-Parallel-model.md:248]:
```suggestion
- Inside the if and else blocks the same field cannot be written to **and** read with an offset in the `I`,`J` parallel axes (order does not matter).
```
Actually, we never say that the I,J horizonal axes are parallel and K is not. Should we extend the domain description at the beginning of the document?

File: GTScript-Parallel-model.md [hunk 3]
```diff
@@ -134,17 +180,119 @@ The computation domain of every statement is extended to ensure that any require
...
+        a = 1
+        b = 2
+    else:
+        a = 2
+        b = 1
+```
+
+translates to:
+
+```python
```
Comment (twicki) [inline GTScript-Parallel-model.md:222]: very clean

Comment (havogt) [reply GTScript-Parallel-model.md:222]: I like that you like your work. ;)

File: GTScript-Parallel-model.md [hunk 4]
```diff
@@ -18,91 +17,139 @@ Trivia: GTScript is an embedded DSL in Python, therefore language syntax is rest
...
 
 The effect of the program is as if statements are executed as follows:
 
-- _computations_ are executed sequentially in the order they appear in the code,
-- vertical _intervals_ are executed sequentially in the order defined by the _iteration policy_ of the _computation_
-- every vertical _interval_ is executed as a sequential for-loop over the `K`-range following the order defined by the iteration policy,
-- every _statement_ inside the _interval_ is executed as a parallel for-loop over the horizontal dimension(s) with no guarantee on the order.
+1. _computations_ are executed sequentially in the order they appear in the code,
+2. vertical _intervals_ are executed sequentially in the order defined by the _iteration policy_ of the _computation_
+3. every vertical _interval_ is executed as a sequential for-loop over the `K`-range following the order defined by the iteration policy,
```
[the remark this answers is not in the document]
Comment (twicki) [reply GTScript-Parallel-model.md:42]: whoops, sorry. I read over that. Ignore this

**Review by egparedes — APPROVED** [review]

**Comment by egparedes** [comment]
from @twicki

I am happy with the proposal. There is one very minor concern on restrictions for field conditionals:
If you have:
```python
if field[1,0,0] > 0:
   field = 1
```

This would translate into
```python
for k:
  parfor ij:
    mask[i,j,k] = field[i+1,j,k]
  parfor ij:
    if mask[i,j,k]:
      field[i+1,j,k] = 1
```
which is legal code that is "versioned" on the conditional whereas
```python
if field[0,0,1] > 0:
  field = 1
```
would translate into
```python
for k:
  parfor ij:
    mask[i,j,k] = field[i,j,k+1]
  parfor ij:
    if mask[i,j,k]:
      field[i+1,j,k] = 1
```
which is not versioned since we put the computation of the conditional inside the same k-loop.
I don't know if we want consistency here but it might be hard to explain the current behaviour.

I see three options, two of which are viable:
- put the mask in it's own k-loop so the mask is always versioned
- keep the current behaviour but add a comment in the proposal.
- disallow off center writes inside condtionals. The probelm with that is that we actually have a real use-case for this in the remapping so I would not do that. But our use case does not write the field so it does not really matter what approach we take

**Comment by egparedes** [comment]
I don't fully understand the examples, I think there are some typos there. My guess about what you mean is:
> ```python
> if field[1,0,0] > 0:
>    field = 1
> ```
> This would translate into
> ```python
> for k:
>   parfor ij:
>     mask[i,j,k] = field[i+1,j,k]
>   parfor ij:
>     if mask[i,j,k]:
>       field[i+1,j,k] = 1
> ```

This should be actually translated as:
```python
for k:
  parfor ij:
    mask[i,j,k] = (field[i+1,j,k] > 0)
  parfor ij:
    if mask[i,j,k]:
      field[i,j,k] = 1
```

and the other case:
> ```python
> if field[0,0,1] > 0:
>   field = 1
> ```
> 
> would translate into
> 
> ```python
> for k:
>   parfor ij:
>     mask[i,j,k] = field[i,j,k+1]
>   parfor ij:
>     if mask[i,j,k]:
>       field[i+1,j,k] = 1
> ```
 
should be actually translated (assuming a non-parallel k iteration) as:
```python
for k:
  parfor ij:
    mask[i,j,k] = (field[i,j,k+1] > 0)
  parfor ij:
    if mask[i,j,k]:
      field[i,j,k] = 1
```
But in this case there isn't any off-center writes, so maybe you meant:
```python
if field[0,0,1] > 0:
  field[0,0,1] = 1
```
which should be translated as:
```python
for k:
  parfor ij:
    mask[i,j,k] = (field[i,j,k+1] > 0)
  parfor ij:
    if mask[i,j,k]:
      field[i,j,k+1] = 1
```
but I still don't see a problem. Could you expand the examples and explain the problem in more detail?

**Comment by twicki** [comment]
The main question I had is if we want to guarantee that the condition is checked before the body is executed. As of this proposal, the condition is pre-computed for every k-level before the body is executed but they are executed inside the k-loop.
In all the cases we've seen this has no implications but it might lead to minor differences in the vertical and the horizontal which I find acceptable but just want to point it out here.

Basically the "versioning" of the field happening in the horizontal allows us to write something like this:
```
if field[-1, 0, 0] > 0:
  field[0, 0, 0] = 1
```
because we are pre-computing the conditional and therefore the body does not affect the condition.
The translation would be 
```
for k:
  parfor ij:
    mask[i, j, k] = (field[i - 1, j, k] > 0)
  parfor ij:
    if mask[i, j, k]:
      field[i, j, k] = 1
```

But if we write the same code with a vertical offset (in a sequential region)
```
if field[0, 0, -1] > 0:
  field[0, 0, 0] = 1
```
the body affects the if-condition on the next level
```
for k:
  parfor ij:
    mask[i, j, k] = (field[i, j, k - 1] > 0)
  parfor ij:
    if mask[i, j, k]:
      field[i, j, k] = 1
```

Generally I think it is fine that the vertical behaves slightly differently than the horizontal but I think part of the effort here is to make sure these nuances are better understood by making the front end different for these cases(i.e disallowing self-assignment in the horizontal versus solver pattern in the vertical).

**Comment by egparedes** [comment]
Thanks for the clarification, I see your point now. I think we can keep the current behavior since it's not ambiguous, but we can improve the PR modifying the pseudo-code for the condition evaluation, because it should be a 2D field evaluated only on the current slice in the `K` loop (which would also help to understand better corner cases like this example).

Of course, the problem of similar syntax meaning different things depending on the axis where is applied still exists, but it's hard to fix. Honestly, the only solution I can think of right now implies to define a different indexing syntax for parallel (IJ) and sequential (K) axes (e.g. `field[0, 1, [-1]]`), but I think it would be highly controversial and not everyone would agree on it...

**Pull request merged**